### DISCUSSION TITLE
Can I use Airflow(Jinja) templated variables in the queue parameter of BaseOperator?

### DISCUSSION THREAD

**Discussion by MattiaGallegati** [apache/airflow#29899] [Q&A] [answered]
Hello everyone,

I'm trying to use a variable of Airflow by calling the Jinja replace mechanism to dynamically decide the target queue.
I simply define the variable and use it from a BashOperator (or any other operator) like:

queue="{{var.value.get('queue')}}"

But it seems it's not working. I've tryed different combination of ' " with or without blank spaces..

I have a distributed architecture with CeleryExecutor.
Is there something I'm missing here?

**Comment by potiuk** [comment, marked as answer]
It's impossible. Jinja parameters get resolved when tasks is about to run - on the target worker. And queue determines where to run the task.
So you have chicken-and-egg problem as you need to know queue value befor running Jinja template rendering.

**Reply by josh-fell** [reply]
If fact, there will also be an explicit check coming to enforce trying to template attributes needed specifically by the BaseOperator: https://github.com/apache/airflow/pull/29821

**Comment by MattiaGallegati** [comment]
Thank you, I discovered exactly that by exploring the Operator class.
By the way,
I found a workaround for the same purpose which was queue=Variabile.get('queue').
This is working as i expected.

We can then say that Variabiles (accessed by Python) and Jinja replacement act in two different moments I suppose.

**Reply by potiuk** [reply]
Yes. Using variable this way is done during parsing the python code not executing the task. This happens in DAG file processoe not in worker. And is not very dynamic because it uses the variable at the moment  the DAG is refreshed by the dag file processor.

**Reply by MattiaGallegati** [reply]
Yes, it perfectly fit to my use case.

Is it the same also for top level python code (e.g. outside the DAG() constructor)?

**Reply by potiuk** [reply]
Yes.

**Reply by potiuk** [reply]
BTW. you should generally avoid running any long operations there - using variables in the way you did is not really recommended https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code - there are better strategies described in https://airflow.apache.org/docs/apache-airflow/stable/howto/dynamic-dag-generation.html#dynamic-dag-generation

**Reply by MattiaGallegati** [reply]
Considering that my "queue" Variable is somehow static, by that I means that will be loaded once and not changed, is there a specific pitfall or risk in using it in my Operator queue parameter like?

BashOperator(
... 
queue=Variabile.get('queue') 
)

**Reply by potiuk** [reply]
https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html#top-level-python-code  explains the risks.
